sys/proc.c

```diff
@@ -53,4 +63,142 @@ proc_t *proc_find(pid_t pid) {
   return p;
 }
 
+void proc_exit(int exitstatus) {
+  /* NOTE: This is a significantly simplified implementation! We currently
+     assume there is only one thread in a process. */
+
+  thread_t *td = thread_self();
+  proc_t *p = td->td_proc;
+  assert(p);
+
+  mtx_lock(&p->p_lock);
+
+  assert(p->p_nthreads == 1);
+  /* If the process had other threads, we'd need to wake the sleeping ones,
+     request all of them except this one to call thread_exit from
+     exc_before_leave (using a TDF_? flag), join all of them to wait until they
+     terminate. */
+
+  /* XXX: If this process has any unwaited zombie children, assign them for
+     adoption by pid1 (init), who will wait for them. */
+
+  /* Clean up process resources. */
+  vm_map_delete(p->p_uspace);
+  fdtab_release(p->p_fdtable);
+
+  /* Record some process statistics that will stay maintained in zombie
+     state. */
+  p->p_exitstatus = exitstatus;
+
+  /* Turn the process into a zombie. */
+  mtx_lock(&all_proc_list_mtx);
+  TAILQ_REMOVE(&all_proc_list, p, p_all);
+  mtx_unlock(&all_proc_list_mtx);
+
+  p->p_state = PRS_ZOMBIE;
+
+  mtx_lock(&zombie_proc_list_mtx);
+  TAILQ_INSERT_TAIL(&zombie_proc_list, p, p_zombie);
+  mtx_unlock(&zombie_proc_list_mtx);
+
+  /* Notify parent possibly waiting for this process to become zombie that it is
+     now. */
+  if (p->p_parent)
+    sleepq_broadcast(&p->p_parent->p_children);
+
+  mtx_unlock(&p->p_lock);
+
+  /* This thread is the last one in the process to exit. */
+  thread_exit();
+}
+
+/* These functions aren't very useful, but they clean up code layout by
+   splitting multiple levels of nested loops */
+static proc_t *proc_find_child(proc_t *p, pid_t pid) {
+  assert(mtx_owned(&p->p_lock));
+  proc_t *child;
+  TAILQ_FOREACH (child, &p->p_children, p_child) {
+    if (child->p_pid == pid)
+      return child;
+  }
+  return NULL;
+}
+
+/* Note: Returned child is locked. */
+static proc_t *proc_find_zombiechild(proc_t *p) {
+  assert(mtx_owned(&p->p_lock));
+  proc_t *child;
+  TAILQ_FOREACH (child, &p->p_children, p_child) {
+    mtx_lock(&child->p_lock);
+    if (child->p_state == PRS_ZOMBIE)
+      return child;
+    mtx_unlock(&child->p_lock);
+  }
+  return NULL;
+}
+
+int do_waitpid(pid_t pid, int *status, int options) {
```

Comment:
I think #324 (Python-like with statement for mutex locking) wouldn't help much in this function - because I need to unlock the mutex before `sleepq_wait` and then re-lock it again. But I have another idea how I could improve these locks by using two wait channels instead of one... I'll commit it soon.